File: scripts/capture_homepage.py
"""
Homepage Visual Design Capture Script
Uses Playwright to capture high-fidelity full-page and sequential sectional screenshots
of the website's homepage, ensuring fonts, lazy-loaded images, and animations are fully rendered.
"""

import os
import sys
import time
import argparse
from pathlib import Path

# Ensure UTF-8 output on Windows consoles
if sys.platform == "win32":
    try:
        sys.stdout.reconfigure(encoding="utf-8")
        sys.stderr.reconfigure(encoding="utf-8")
    except Exception:
        pass

from playwright.sync_api import sync_playwright

PROJECT_ROOT = Path(__file__).resolve().parent.parent
DEFAULT_OUTPUT_DIR = PROJECT_ROOT / "screenshots" / "homepage"


def wait_for_page_ready(page, settle_delay_ms: int = 1500):
    """
    Ensures that fonts, lazy-loaded images, dynamic UI elements,
    and layout shifts are completely loaded and settled.
    """
    print("[PAGE] Waiting for web fonts to load (document.fonts.ready)...")
    try:
        page.evaluate("() => document.fonts.ready")
    except Exception as e:
        print(f"   [NOTICE] Font check warning: {e}")

    print("[PAGE] Scrolling through page to trigger lazy loading & intersection observers...")
    page.evaluate("""async () => {
        const step = window.innerHeight * 0.75;
        const totalHeight = Math.max(
            document.body.scrollHeight,
            document.documentElement.scrollHeight
        );
        for (let pos = 0; pos < totalHeight; pos += step) {
            window.scrollTo({ top: pos, behavior: 'instant' });
            await new Promise(r => setTimeout(r, 100));
        }
        window.scrollTo({ top: totalHeight, behavior: 'instant' });
        await new Promise(r => setTimeout(r, 150));
    }""")

    print("[PAGE] Waiting for all images to complete loading...")
    page.evaluate("""async () => {
        const images = Array.from(document.querySelectorAll('img'));
        await Promise.all(images.map(img => {
            if (img.complete) return Promise.resolve();
            return new Promise((resolve) => {
                img.addEventListener('load', resolve, { once: true });
                img.addEventListener('error', resolve, { once: true });
            });
        }));
    }""")

    print(f"[PAGE] Settling animations and dynamic elements ({settle_delay_ms}ms)...")
    time.sleep(settle_delay_ms / 1000.0)

    # Scroll back to top
    page.evaluate("window.scrollTo({ top: 0, behavior: 'instant' })")
    time.sleep(0.3)


def capture_screenshots(
    url: str,
    output_dir: Path,
    mode: str = "all",
    sections_count: int = 4,
    viewport_width: int = 1440,
    viewport_height: int = 900,
    scale_factor: int = 2,
    theme: str = "dark",
    settle_delay_ms: int = 1500,
    headful: bool = False
):
    output_dir.mkdir(parents=True, exist_ok=True)
    captured_files = []

    print("=" * 60)
    print(" FreeIQExam Homepage Visual Design Capture")
    print("=" * 60)
    print(f"Target URL:        {url}")
    print(f"Output Directory:  {output_dir}")
    print(f"Mode:              {mode}")
    print(f"Viewport:          {viewport_width}x{viewport_height} (Scale: {scale_factor}x -> {viewport_width * scale_factor}x{viewport_height * scale_factor})")
    print(f"Theme / Scheme:    {theme}")
    print("=" * 60)

    with sync_playwright() as p:
        browser = p.chromium.launch(
            headless=not headful,
        )

        context = browser.new_context(
            viewport={"width": viewport_width, "height": viewport_height},
            device_scale_factor=scale_factor,
            color_scheme=theme if theme in ["light", "dark"] else "no-preference"
        )

        page = context.new_page()

        print(f"\n[NAV] Navigating to {url}...")
        try:
            page.goto(url, wait_until="networkidle", timeout=20000)
        except Exception as e:
            print(f"   [NOTICE] Navigation wait condition notice ({e}). Continuing with loaded DOM...")
            page.wait_for_load_state("load")

        # Wait for all assets, fonts, dynamic elements
        wait_for_page_ready(page, settle_delay_ms=settle_delay_ms)

        page_title = page.title()
        print(f"[PAGE] Page Title: \"{page_title}\"")

        # Measure page height
        scroll_height = page.evaluate("() => Math.max(document.body.scrollHeight, document.documentElement.scrollHeight)")
        print(f"[PAGE] Total Page Scroll Height: {scroll_height}px")

        timestamp = time.strftime("%Y%m%d_%H%M%S")

        # 1. Full Page Screenshot
        if mode in ("all", "fullpage"):
            fullpage_filename = f"homepage_{theme}_fullpage_{timestamp}.png"
            fullpage_path = output_dir / fullpage_filename
            print(f"\n[CAPTURE] Capturing full-page screenshot: {fullpage_filename}...")
            page.evaluate("window.scrollTo(0, 0)")
            time.sleep(0.3)
            page.screenshot(path=str(fullpage_path), full_page=True)
            size_kb = os.path.getsize(fullpage_path) / 1024
            print(f"   [SAVED] {fullpage_path} ({size_kb:.1f} KB)")
            captured_files.append(fullpage_path)

        # 2. Sequential Sectional Screenshots
        if mode in ("all", "sections"):
            print(f"\n[CAPTURE] Capturing {sections_count} sequential sectional screenshots (top to bottom)...")
            max_scroll = max(0, scroll_height - viewport_height)

            section_labels = [
                "top_hero",
                "upper_middle_features",
                "lower_middle_details",
                "bottom_footer"
            ]

            for i in range(sections_count):
                if sections_count == 1:
                    scroll_y = 0
                else:
                    scroll_y = int((i / (sections_count - 1)) * max_scroll)

                # Assign descriptive label
                if i < len(section_labels) and sections_count == len(section_labels):
                    label = section_labels[i]
                elif i == 0:
                    label = "top_hero"
                elif i == sections_count - 1:
                    label = "bottom_footer"
                else:
                    label = f"middle_section_{i + 1}"

                section_filename = f"homepage_{theme}_part{i+1}_{label}_{timestamp}.png"
                section_path = output_dir / section_filename

                # Scroll to target position
                page.evaluate(f"window.scrollTo({{ top: {scroll_y}, behavior: 'instant' }})")
                time.sleep(0.4)

                page.screenshot(path=str(section_path), full_page=False)
                size_kb = os.path.getsize(section_path) / 1024
                print(f"   [SAVED] Part {i+1}/{sections_count} (scroll: {scroll_y}px / {scroll_height}px) -> {section_filename} ({size_kb:.1f} KB)")
                captured_files.append(section_path)

        context.close()
        browser.close()

    print("\n" + "=" * 60)
    print(f"[COMPLETE] Successfully captured {len(captured_files)} screenshot(s)!")
    print(f"[LOCATION] {output_dir}")
    print("=" * 60)
    return captured_files


def main():
    parser = argparse.ArgumentParser(
        description="Capture full-page and sectional screenshots of FreeIQExam homepage."
    )
    parser.add_argument(
        "--url",
        default="http://localhost:4321",
        help="URL of the homepage to capture (default: http://localhost:4321)"
    )
    parser.add_argument(
        "--output-dir",
        default=str(DEFAULT_OUTPUT_DIR),
        help=f"Output directory for screenshots (default: {DEFAULT_OUTPUT_DIR})"
    )
    parser.add_argument(
        "--mode",
        choices=["all", "fullpage", "sections"],
        default="all",
        help="Screenshot mode: 'all' (default, both fullpage and sections), 'fullpage', or 'sections'"
    )
    parser.add_argument(
        "--sections",
        type=int,
        default=4,
        help="Number of sequential sectional screenshots to take (default: 4)"
    )
    parser.add_argument(
        "--width",
        type=int,
        default=1440,
        help="Viewport width in pixels (default: 1440)"
    )
    parser.add_argument(
        "--height",
        type=int,
        default=900,
        help="Viewport height in pixels (default: 900)"
    )
    parser.add_argument(
        "--scale",
        type=int,
        default=2,
        help="Device pixel scale factor for crisp retina screenshots (default: 2)"
    )
    parser.add_argument(
        "--theme",
        choices=["dark", "light", "both"],
        default="dark",
        help="Color scheme preference: 'dark', 'light', or 'both' (default: dark)"
    )
    parser.add_argument(
        "--settle-delay",
        type=int,
        default=1500,
        help="Extra settling delay in milliseconds after scrolling (default: 1500)"
    )
    parser.add_argument(
        "--headful",
        action="store_true",
        help="Run browser with visible GUI window instead of headless"
    )

    args = parser.parse_args()
    out_dir = Path(args.output_dir)

    themes = ["dark", "light"] if args.theme == "both" else [args.theme]
    all_captured = []
    for t in themes:
        captured = capture_screenshots(
            url=args.url,
            output_dir=out_dir,
            mode=args.mode,
            sections_count=args.sections,
            viewport_width=args.width,
            viewport_height=args.height,
            scale_factor=args.scale,
            theme=t,
            settle_delay_ms=args.settle_delay,
            headful=args.headful
        )
        all_captured.extend(captured)


if __name__ == "__main__":
    main()

File: scripts/capture_homepage.mjs
/**
 * FreeIQExam Homepage Visual Capture (Node.js / Playwright)
 * Captures full-page and sequential sectional screenshots of the homepage.
 *
 * Usage:
 *   node scripts/capture_homepage.mjs [--url http://localhost:4321] [--mode all]
 *   npx --package=playwright node scripts/capture_homepage.mjs
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { parseArgs } from 'node:util';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const PROJECT_ROOT = path.resolve(__dirname, '..');
const DEFAULT_OUTPUT_DIR = path.join(PROJECT_ROOT, 'screenshots', 'homepage');

// Parse CLI flags
const { values: args } = parseArgs({
  options: {
    url: { type: 'string', default: 'http://localhost:4321' },
    'output-dir': { type: 'string', default: DEFAULT_OUTPUT_DIR },
    mode: { type: 'string', default: 'all' }, // all, fullpage, sections
    sections: { type: 'string', default: '4' },
    width: { type: 'string', default: '1440' },
    height: { type: 'string', default: '900' },
    scale: { type: 'string', default: '2' },
    theme: { type: 'string', default: 'dark' }, // dark, light, both
    'settle-delay': { type: 'string', default: '1500' },
    help: { type: 'boolean', default: false }
  },
  allowPositionals: true
});

if (args.help) {
  console.log(`
FreeIQExam Homepage Visual Capture
Options:
  --url <url>             URL to capture (default: http://localhost:4321)
  --output-dir <path>     Directory to save screenshots (default: screenshots/homepage)
  --mode <mode>           "all", "fullpage", or "sections" (default: "all")
  --sections <n>          Number of sequential sectional slices (default: 4)
  --width <pixels>        Viewport width (default: 1440)
  --height <pixels>       Viewport height (default: 900)
  --scale <factor>        Device scale factor (default: 2)
  --theme <theme>         "dark", "light", or "both" (default: "dark")
  --settle-delay <ms>     Delay in ms to let animations settle (default: 1500)
  `);
  process.exit(0);
}

let playwright;
try {
  playwright = await import('playwright');
} catch {
  console.error(`
[ERROR] Playwright is not installed in node_modules.
You can run:
  npm install -D playwright
or run via npx:
  npx --package=playwright node scripts/capture_homepage.mjs
or run the included Python script:
  python scripts/capture_homepage.py
  `);
  process.exit(1);
}

const { chromium } = playwright;

async function waitForPageReady(page, settleDelayMs) {
  console.log('[PAGE] Waiting for web fonts to load (document.fonts.ready)...');
  try {
    await page.evaluate(() => document.fonts.ready);
  } catch (e) {
    console.log('   [NOTICE] Font readiness check notice:', e.message);
  }

  console.log('[PAGE] Scrolling through page to trigger lazy loading & animations...');
  await page.evaluate(async () => {
    const step = window.innerHeight * 0.75;
    const totalHeight = Math.max(document.body.scrollHeight, document.documentElement.scrollHeight);
    for (let pos = 0; pos < totalHeight; pos += step) {
      window.scrollTo({ top: pos, behavior: 'instant' });
      await new Promise((r) => setTimeout(r, 100));
    }
    window.scrollTo({ top: totalHeight, behavior: 'instant' });
    await new Promise((r) => setTimeout(r, 150));
  });

  console.log('[PAGE] Waiting for all images to complete loading...');
  await page.evaluate(async () => {
    const images = Array.from(document.querySelectorAll('img'));
    await Promise.all(
      images.map((img) => {
        if (img.complete) return Promise.resolve();
        return new Promise((resolve) => {
          img.addEventListener('load', resolve, { once: true });
          img.addEventListener('error', resolve, { once: true });
        });
      })
    );
  });

  console.log(`[PAGE] Settling animations and dynamic elements (${settleDelayMs}ms)...`);
  await page.waitForTimeout(settleDelayMs);

  await page.evaluate(() => window.scrollTo({ top: 0, behavior: 'instant' }));
  await page.waitForTimeout(300);
}

async function capture(theme) {
  const outputDir = path.resolve(args['output-dir']);
  fs.mkdirSync(outputDir, { recursive: true });

  const viewportWidth = parseInt(args.width, 10);
  const viewportHeight = parseInt(args.height, 10);
  const scale = parseInt(args.scale, 10);
  const sectionsCount = parseInt(args.sections, 10);
  const settleDelay = parseInt(args['settle-delay'], 10);

  console.log('='.repeat(60));
  console.log(' FreeIQExam Homepage Visual Capture');
  console.log('='.repeat(60));
  console.log(`Target URL:        ${args.url}`);
  console.log(`Output Directory:  ${outputDir}`);
  console.log(`Mode:              ${args.mode}`);
  console.log(`Viewport:          ${viewportWidth}x${viewportHeight} (Scale: ${scale}x)`);
  console.log(`Theme:             ${theme}`);
  console.log('='.repeat(60));

  const browser = await chromium.launch();
  const context = await browser.newContext({
    viewport: { width: viewportWidth, height: viewportHeight },
    deviceScaleFactor: scale,
    colorScheme: theme === 'light' || theme === 'dark' ? theme : 'no-preference'
  });

  const page = await context.newPage();

  console.log(`\n[NAV] Navigating to ${args.url}...`);
  try {
    await page.goto(args.url, { waitUntil: 'networkidle', timeout: 20000 });
  } catch (e) {
    console.log('   [NOTICE] Networkidle timeout reached, proceeding with loaded page...');
    await page.waitForLoadState('load');
  }

  await waitForPageReady(page, settleDelay);

  const pageTitle = await page.title();
  console.log(`[PAGE] Page Title: "${pageTitle}"`);

  const scrollHeight = await page.evaluate(() =>
    Math.max(document.body.scrollHeight, document.documentElement.scrollHeight)
  );
  console.log(`[PAGE] Total Page Scroll Height: ${scrollHeight}px`);

  const now = new Date();
  const timestamp = now.toISOString().replace(/[-:T]/g, '').slice(0, 14);

  // 1. Full Page Screenshot
  if (args.mode === 'all' || args.mode === 'fullpage') {
    const fullFileName = `homepage_${theme}_fullpage_${timestamp}.png`;
    const fullFilePath = path.join(outputDir, fullFileName);
    console.log(`\n[CAPTURE] Capturing full-page screenshot: ${fullFileName}...`);
    await page.evaluate(() => window.scrollTo(0, 0));
    await page.waitForTimeout(300);
    await page.screenshot({ path: fullFilePath, fullPage: true });
    const sizeKb = (fs.statSync(fullFilePath).size / 1024).toFixed(1);
    console.log(`   [SAVED] ${fullFilePath} (${sizeKb} KB)`);
  }

  // 2. Sequential Sectional Screenshots
  if (args.mode === 'all' || args.mode === 'sections') {
    console.log(`\n[CAPTURE] Capturing ${sectionsCount} sequential sectional screenshots (top to bottom)...`);
    const maxScroll = Math.max(0, scrollHeight - viewportHeight);
    const sectionLabels = [
      'top_hero',
      'upper_middle_features',
      'lower_middle_details',
      'bottom_footer'
    ];

    for (let i = 0; i < sectionsCount; i++) {
      const scrollY = sectionsCount === 1 ? 0 : Math.round((i / (sectionsCount - 1)) * maxScroll);
      let label = `section_${i + 1}`;
      if (i < sectionLabels.length && sectionsCount === sectionLabels.length) {
        label = sectionLabels[i];
      } else if (i === 0) {
        label = 'top_hero';
      } else if (i === sectionsCount - 1) {
        label = 'bottom_footer';
      }

      const sectionFileName = `homepage_${theme}_part${i + 1}_${label}_${timestamp}.png`;
      const sectionFilePath = path.join(outputDir, sectionFileName);

      await page.evaluate((y) => window.scrollTo({ top: y, behavior: 'instant' }), scrollY);
      await page.waitForTimeout(400);

      await page.screenshot({ path: sectionFilePath, fullPage: false });
      const sizeKb = (fs.statSync(sectionFilePath).size / 1024).toFixed(1);
      console.log(`   [SAVED] Part ${i + 1}/${sectionsCount} (scroll: ${scrollY}px / ${scrollHeight}px) -> ${sectionFileName} (${sizeKb} KB)`);
    }
  }

  await context.close();
  await browser.close();

  console.log('\n' + '='.repeat(60));
  console.log(`[COMPLETE] Screenshots saved to: ${outputDir}`);
  console.log('='.repeat(60));
}

const themes = args.theme === 'both' ? ['dark', 'light'] : [args.theme];
for (const t of themes) {
  await capture(t);
}
